Allow LN tail to have an origin of `TopCentre`

longnguyen2004 opened the pull request:
Originally discussed at https://github.com/ppy/osu/discussions/26470

This PR adds a new `skin.ini` setting `HoldNoteTailOrigin`, with a value of 0 for Bottom and 1 for Top, which allows the user to adjust the origin of the tail piece, avoiding the huge texture hack found in current skins.
(Should the setting be an array instead? Each column is independent, so theoretically it should be an array, but I'm not aware of any skins that mix and match different textures for each column) 

Demo
<video src="https://github.com/ppy/osu/assets/15729831/353dd119-9558-4397-874c-0312530389f1"></video>
(Ignore the overlapped part, I didn't blank out the bottom half of the tail)

longnguyen2004 committed:
Add `HoldNoteTailOrigin` settings

longnguyen2004 committed:
Make the setting works

longnguyen2004 committed:
Fix InspectCode

longnguyen2004 committed:
Properly reset the tail origin on skin change

longnguyen2004 committed:
Merge branch 'master' into hold-note-tail-topcentre

longnguyen2004 committed:
Fix comment

longnguyen2004 committed:
Merge branch 'hold-note-tail-topcentre' of https://github.com/longnguyen2004/osu into hold-note-tail-topcentre

frenzibyte commented on osu.Game.Rulesets.Mania/Objects/Drawables/DrawableHoldNote.cs:
```diff
@@ -260,6 +272,15 @@ protected override void Update()
+        protected override void ApplySkin(ISkinSource skin, bool allowFallback)
+        {
+            base.ApplySkin(skin, allowFallback);
+            var skinTailOrigin = skin.GetConfig<ManiaSkinConfigurationLookup, Anchor>(
+                new ManiaSkinConfigurationLookup(LegacyManiaSkinConfigurationLookups.HoldNoteTailOrigin)
+            );
+            tailOrigin = skinTailOrigin ?? new Bindable<Anchor>(Anchor.BottomCentre);
```
Don't store bindable references from external sources, also use bindable feedback.

```diff
diff --git a/osu.Game.Rulesets.Mania/Objects/Drawables/DrawableHoldNote.cs b/osu.Game.Rulesets.Mania/Objects/Drawables/DrawableHoldNote.cs
index 04b09e9d0f..14e31a5568 100644
--- a/osu.Game.Rulesets.Mania/Objects/Drawables/DrawableHoldNote.cs
+++ b/osu.Game.Rulesets.Mania/Objects/Drawables/DrawableHoldNote.cs
@@ -32,9 +32,10 @@ public partial class DrawableHoldNote : DrawableManiaHitObject<HoldNote>, IKeyBi
 
         public IBindable<bool> IsHitting => isHitting;
 
-        private IBindable<Anchor> tailOrigin;
         private readonly Bindable<bool> isHitting = new Bindable<bool>();
 
+        private readonly Bindable<Anchor> tailOrigin = new Bindable<Anchor>(Anchor.BottomCentre);
+
         public DrawableHoldNoteHead Head => headContainer.Child;
         public DrawableHoldNoteTail Tail => tailContainer.Child;
         public DrawableHoldNoteBody Body => bodyContainer.Child;
@@ -129,6 +130,10 @@ protected override void LoadComplete()
             base.LoadComplete();
 
             isHitting.BindValueChanged(updateSlidingSample, true);
+
+            tailOrigin.BindValueChanged(_ => updateTailOrigin());
+            Direction.BindValueChanged(_ => updateTailOrigin());
+            updateTailOrigin();
         }
 
         protected override void OnApply()
@@ -242,17 +247,12 @@ protected override void Update()
             // The rationale for this is account for heads/tails with corner radius.
             bodyPiece.Y = (Direction.Value == ScrollingDirection.Up ? 1 : -1) * Head.Height / 2;
             bodyPiece.Height = DrawHeight - Head.Height / 2;
+
             if (tailOrigin.Value == Anchor.TopCentre)
                 bodyPiece.Height -= Tail.Height / 2;
             else
                 bodyPiece.Height += Tail.Height / 2;
 
-            // Update the origin of the tail piece, taking into account the scrolling direction
-            if (Direction.Value == ScrollingDirection.Up)
-                Tail.Origin = tailOrigin.Value == Anchor.TopCentre ? Anchor.BottomCentre : Anchor.TopCentre;
-            else
-                Tail.Origin = tailOrigin.Value;
-
             if (Time.Current >= HitObject.StartTime)
             {
                 // As the note is being held, adjust the size of the sizing container. This has two effects:
@@ -275,10 +275,7 @@ protected override void Update()
         protected override void ApplySkin(ISkinSource skin, bool allowFallback)
         {
             base.ApplySkin(skin, allowFallback);
-            var skinTailOrigin = skin.GetConfig<ManiaSkinConfigurationLookup, Anchor>(
-                new ManiaSkinConfigurationLookup(LegacyManiaSkinConfigurationLookups.HoldNoteTailOrigin)
-            );
-            tailOrigin = skinTailOrigin ?? new Bindable<Anchor>(Anchor.BottomCentre);
+            tailOrigin.Value = skin.GetConfig<ManiaSkinConfigurationLookup, Anchor>(new ManiaSkinConfigurationLookup(LegacyManiaSkinConfigurationLookups.HoldNoteTailOrigin))?.Value ?? Anchor.BottomCentre;
         }
 
         protected override void CheckForResult(bool userTriggered, double timeOffset)
@@ -392,6 +389,14 @@ private void updateSlidingSample(ValueChangedEvent<bool> tracking)
                 slidingSample?.Stop();
         }
 
+        private void updateTailOrigin()
+        {
+            if (Direction.Value == ScrollingDirection.Up)
+                Tail.Origin = tailOrigin.Value == Anchor.TopCentre ? Anchor.BottomCentre : Anchor.TopCentre;
+            else
+                Tail.Origin = tailOrigin.Value;
+        }
+
         protected override void OnFree()
         {
             slidingSample.ClearSamples();
```

longnguyen2004 replied:
I found out it was due to https://github.com/ppy/osu/blob/4490dbf8960214dcc9d3f67dbfc88ab5255c6e94/osu.Game.Rulesets.Mania/Objects/Drawables/DrawableManiaHitObject.cs#L68-L71

To workaround this, I'd have to override `OnDirectionChanged` inside `DrawableHoldNoteTail` and duplicate the lookup logic, which I'd like to avoid. Is there another way to do this? Maybe make `tailOrigin` public and reference that inside the tail class?

longnguyen2004 committed:
Rename key + Create enum for values

longnguyen2004 commented on osu.Game.Rulesets.Mania/Objects/Drawables/DrawableHoldNote.cs:
```diff
@@ -239,7 +241,17 @@ protected override void Update()
+            else
+                bodyPiece.Height += Tail.Height / 2;
+
+            // Update the origin of the tail piece, taking into account the scrolling direction
+            if (Direction.Value == ScrollingDirection.Up)
+                Tail.Origin = tailOrigin.Value == Anchor.TopCentre ? Anchor.BottomCentre : Anchor.TopCentre;
+            else
+                Tail.Origin = tailOrigin.Value;
```
It doesn't work without being in `Update` for whatever reason. I can see the origin being updated to `TopCentre` in the debugger, yet the value shown in the Draw Visualizer is `BottomCentre` anyway.

![image](https://github.com/ppy/osu/assets/15729831/1a3ede52-0c80-41df-9fb2-a60d977f6ae5)
![image](https://github.com/ppy/osu/assets/15729831/01f2f223-9c53-4874-973d-221414d59977)

I've also noticed stuff like the container's padding and `bodyPiece.Y` being in update even when they don't change, so I thought it was fine to put the tail's origin there.

longnguyen2004 commented:
Skin for testing (rename to osk): [R Skin v3.1 (7k Fade LN).zip](https://github.com/ppy/osu/files/14772607/R.Skin.v3.1.7k.Fade.LN.zip)
Map for testing: https://osu.ppy.sh/beatmapsets/861816#mania/1802831

longnguyen2004 committed:
Apply frenzibyte's patch with changes

longnguyen2004 committed:
Fix InspectCode

frenzibyte commented:
> Also, is there a chance that this will get backported to stable, like the new `songselect-top` and `songselect-bottom` element?

That's off-topic in this PR, stable is generally feature-locked so it's unlikely that such a feature will be backported, but that should not be discussed in here anyway.

frenzibyte committed:
Reorder and cleanup code

frenzibyte committed:
Merge branch 'master' into hold-note-tail-topcentre

frenzibyte committed:
Add visual test coverage

frenzibyte committed:
Use direction-agnostic names for tail origin setting

The name `Top` and `Bottom` does not make sense in combination with `ScrollingDirection` which already flips anchor/origin vertically.

frenzibyte reviewed (dismissed):
From a code perspective, this looks acceptable to me after few tidy-ups. However, I'm not confident enough to merge and have this setting exist by myself. I'll leave that to @peppy considering this approach was proposed by him.

longnguyen2004 commented on osu.Game.Rulesets.Mania/Objects/Drawables/DrawableHoldNoteTail.cs:
```diff
@@ -52,5 +66,25 @@ protected override HitResult GetCappedResult(HitResult result)
+
+        protected override void ApplySkin(ISkinSource skin, bool allowFallback)
+        {
+            base.ApplySkin(skin, allowFallback);
+            tailOrigin.Value = skin.GetConfig<ManiaSkinConfigurationLookup, HoldNoteTailOrigin>(new ManiaSkinConfigurationLookup(LegacyManiaSkinConfigurationLookups.HoldNoteTailOrigin))?.Value ?? HoldNoteTailOrigin.Regular;
+        }
+
+        protected override void OnDirectionChanged(ValueChangedEvent<ScrollingDirection> e)
+        {
+            base.OnDirectionChanged(e);
+            updateTailOrigin();
+        }
+
+        private void updateTailOrigin()
+        {
+            if (Direction.Value == ScrollingDirection.Up)
+                Origin = tailOrigin.Value == HoldNoteTailOrigin.Inverted ? Anchor.BottomCentre : Anchor.TopCentre;
+            else
+                Origin = tailOrigin.Value == HoldNoteTailOrigin.Inverted ? Anchor.TopCentre : Anchor.BottomCentre;
+        }
```
Regular variable is probably fine, this is only used in `DrawableHoldNote(Tail)` anyway

longnguyen2004 replied:
I just realized they're used in tests also, will revert it later.

longnguyen2004 committed:
Replace bindable with regular variable

longnguyen2004 committed:
Move enum to `LegacyManiaSkinConfigurationLookup.cs`

longnguyen2004 committed:
Fix tests + InspectCode

peppy commented:
Blocked pending https://github.com/ppy/osu/pull/27773 merge / discussion.

longnguyen2004 commented:
Since #27773 is now scrapped, how should this PR be progressed? Do we wait for the 2nd iteration of the new skinning format?

peppy commented:
> Since #27773 is now scrapped, how should this PR be progressed? Do we wait for the 2nd iteration of the new skinning format?

I'm not sure it's required, but it should arrive this week so stay tuned.

longnguyen2004 committed:
Merge branch 'master' into hold-note-tail-topcentre

longnguyen2004 committed:
Merge branch 'master' into hold-note-tail-topcentre
